README.md
# qrt [![Build Status](https://travis-ci.com/portothree/qrt.svg?branch=master)](https://travis-ci.com/github/portothree/qrt)

> Air Quality Toolkit

## Install

```
$ npm install qrt
```

## Usage

```js
const { windDirectionMean } = require('qrt');

windDirectionMean([360, 45, 90]);
//=> 45
```

## API

### windDirectionMean(...args)

-   args

    -   Wind direction only: `Array<number>`
    -   Wind direction and speed: `Array<number>, Array<number>`
    -   Wind direction and speed in a '2d' array: `<Array<Array<number>>>`

```js
// Wind directions only
windDirectionMean([360, 45, 90]); // => 45

// Wind directions and speeds
windDirectionMean([360, 45, 90], [1.48, 1.78, 3.44]); // => 59.76

// Wind directions and speeds in '2d' format
windDirectionMean([
	[360, 1.48],
	[45, 1.78],
	[90, 3.44],
]); // => 59.76
```

### airQualityIndex(PM10, PM25, NO2, O3, SO2)

-   PM10 - Average hourly as `Number`
-   PM25 - Average hourly as `Number`
-   NO2 - Average hourly as `Number`
-   O3 - Average hourly as `Number`
-   SO2 - Average hourly as `Number`

```js
airQualityIndex(5, 0, 0, 0, 0); // => { AQI_POL: 'PM10', AQI_GLOBAL: 5 }
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.3.0](https://github.com/portothree/qrt/compare/v1.2.0...v1.3.0) (2020-12-11)


### Features

* add Air Quality Index calculation function ([47e4b42](https://github.com/portothree/qrt/commit/47e4b42c87e82e7466df39e7ee8f017efa22b3ac))


test.js
const cases = require('jest-in-case');
const { windDirectionMean, airQualityIndex } = require('./index');

describe('Air Quality Toolkit', () => {
	describe('Wind Direction Mean', () => {
		test('Throws if non array is passed as input', () => {
			expect(() => windDirectionMean('random')).toThrow();
		});
		test('Calculate only directions', () => {
			expect(windDirectionMean([360, 45, 90])).toEqual(45);
		});
		test('Calculate with directions and speed', () => {
			expect(
				windDirectionMean([360, 45, 90], [1.48, 1.78, 3.44])
			).toEqual(59.76384812527999);
		});
		test('Calculate when receive array of arrays with directions and speed', () => {
			expect(
				windDirectionMean([
					[360, 1.48],
					[45, 1.78],
					[90, 3.44],
				])
			).toEqual(59.76384812527999);
		});
	});

	cases(
		'Air Quality Index',
		opts => {
			const [pollutant, aqi] = opts.name.split(':');
			const [PM10, PM25, NO2, O3, SO2] = opts.values;
			const result = airQualityIndex(PM10, PM25, NO2, O3, SO2);

			expect(result).toEqual({
				AQI_POL: pollutant,
				AQI_GLOBAL: Number(aqi),
			});
		},
		[
			{ name: 'PM10:5', values: [1, 0, 0, 0, 0] },
			{ name: 'PM10:5', values: [20, 0, 0, 0, 0] },
			{ name: 'PM10:4', values: [22, 0, 0, 0, 0] },
			{ name: 'PM10:4', values: [35, 0, 0, 0, 0] },
			{ name: 'PM10:3', values: [37, 0, 0, 0, 0] },
			{ name: 'PM10:3', values: [50, 0, 0, 0, 0] },
			{ name: 'PM10:2', values: [52, 0, 0, 0, 0] },
			{ name: 'PM10:2', values: [100, 0, 0, 0, 0] },
			{ name: 'PM10:1', values: [102, 0, 0, 0, 0] },
			{ name: 'PM10:5', values: [0, 1, 0, 0, 0] },
			{ name: 'PM10:5', values: [0, 10, 0, 0, 0] },
			{ name: 'PM25:4', values: [0, 12, 0, 0, 0] },
			{ name: 'PM25:4', values: [0, 20, 0, 0, 0] },
			{ name: 'PM25:3', values: [0, 22, 0, 0, 0] },
			{ name: 'PM25:3', values: [0, 25, 0, 0, 0] },
			{ name: 'PM25:2', values: [0, 27, 0, 0, 0] },
			{ name: 'PM25:2', values: [0, 50, 0, 0, 0] },
			{ name: 'PM25:1', values: [0, 52, 0, 0, 0] },
			{ name: 'PM10:5', values: [0, 0, 1, 0, 0] },
			{ name: 'PM10:5', values: [0, 0, 39, 0, 0] },
			{ name: 'NO2:4', values: [0, 0, 42, 0, 0] },
			{ name: 'NO2:4', values: [0, 0, 100, 0, 0] },
			{ name: 'NO2:3', values: [0, 0, 102, 0, 0] },
			{ name: 'NO2:3', values: [0, 0, 200, 0, 0] },
			{ name: 'NO2:2', values: [0, 0, 202, 0, 0] },
			{ name: 'NO2:2', values: [0, 0, 400, 0, 0] },
			{ name: 'NO2:1', values: [0, 0, 402, 0, 0] },
			{ name: 'PM10:5', values: [0, 0, 0, 0, 1] },
			{ name: 'PM10:5', values: [0, 0, 0, 0, 99] },
			{ name: 'SO2:4', values: [0, 0, 0, 0, 102] },
			{ name: 'SO2:4', values: [0, 0, 0, 0, 200] },
			{ name: 'SO2:3', values: [0, 0, 0, 0, 202] },
			{ name: 'SO2:3', values: [0, 0, 0, 0, 350] },
			{ name: 'SO2:2', values: [0, 0, 0, 0, 352] },
			{ name: 'SO2:2', values: [0, 0, 0, 0, 500] },
			{ name: 'SO2:1', values: [0, 0, 0, 0, 502] },
		]
	);
});


index.js
'use strict';
const { deg2rad, rad2deg } = require('./utils.js');

module.exports = {
	/**
	 * @param {Array} [directions] - array of directions in degrees
	 * @param {Array} [input] - array of arrays containing
	 * wind direction and speed
	 * @param {Array} [speed] - array of speeds in m/s
	 * @return {number}
	 *
	 * @example
	 *
	 *	avgWind([360, 45, 90]);
	 *	avgWind([[360, 1.48], [45, 1.78], [90, 3.44]]);
	 *	avgWind([360, 45, 90], [1.48, 1.78, 3.44]);
	 */
	windDirectionMean: (...args) => {
		let sinSum = 0;
		let cosSum = 0;

		if (args.length === 1) {
			if (!Array.isArray(args[0])) {
				throw new TypeError(`Expected a array, got ${typeof args[0]}`);
			}

			// The argument is an array of arrays
			if (Array.isArray(args[0][0])) {
				args[0].forEach(([direction, speed]) => {
					sinSum += speed * Math.sin(deg2rad(direction));
					cosSum += speed * Math.cos(deg2rad(direction));
				});
			} else {
				// Only an array of directions was passed
				args[0].forEach(direction => {
					sinSum += Math.sin(deg2rad(direction));
					cosSum += Math.cos(deg2rad(direction));
				});
			}
		}

		// Expect first argument to be an array of directions
		// and the second to be of speed
		if (args.length === 2) {
			if (!Array.isArray(args[1])) {
				throw new TypeError(`Expected a array, got ${typeof args[0]}`);
			}

			const directions = args[0];
			const speeds = args[1];

			directions.forEach((direction, index) => {
				const speed = speeds[index];

				sinSum += speed * Math.sin(deg2rad(direction));
				cosSum += speed * Math.cos(deg2rad(direction));
			});
		}

		return (rad2deg(Math.atan2(sinSum, cosSum)) + 360) % 360;
	},

	/**
	 * Air Quality Index
	 */
	airQualityIndex(med_PM10, med_PM25, med_NO2, med_O3, med_SO2) {
		function calcSensorAqi(aqiRules, sensorValue) {
			return Object.entries(aqiRules).reduce(
				(accumulator, [aqi, range]) => {
					const [min, max] = range;

					if (sensorValue > min && sensorValue < max) {
						accumulator = Number(aqi);
					}

					return accumulator;
				},
				5
			);
		}

		// AQI: [min, max]
		const rules = {
			PM10: {
				5: [0, 21],
				4: [21, 36],
				3: [36, 51],
				2: [51, 101],
				1: [101, Infinity],
			},
			PM25: {
				5: [0, 11],
				4: [11, 21],
				3: [21, 26],
				2: [26, 51],
				1: [51, Infinity],
			},
			NO2: {
				5: [0, 40],
				4: [40, 101],
				3: [101, 201],
				2: [201, 401],
				1: [401, Infinity],
			},
			SO2: {
				5: [0, 100],
				4: [101, 201],
				3: [201, 351],
				2: [351, 501],
				1: [501, Infinity],
			},
		};

		const aqi = {
			PM10: calcSensorAqi(rules.PM10, med_PM10),
			PM25: calcSensorAqi(rules.PM25, med_PM25),
			NO2: calcSensorAqi(rules.NO2, med_NO2),
			SO2: calcSensorAqi(rules.SO2, med_SO2),
		};

		const [AQI_POL, AQI_GLOBAL] = Object.entries(aqi).reduce(
			(accumulator, current) => {
				const [accSensor, accValue] = accumulator;
				const [currSensor, currValue] = current;

				if (accValue > currValue) {
					accumulator = current;
				}

				return accumulator;
			},
			['PM10', aqi.PM10]
		);

		return { AQI_POL, AQI_GLOBAL };
	},
};
